### PR TITLE
HEEDLS-853 Fix UserDataService queries

### PR DESCRIPTION
**Pull request by AlexJacksonDS**
### JIRA link
https://softwiretech.atlassian.net/browse/HEEDLS-853

### Description
Provisionally fixed all the queries in UserDataService, some of the unit tests may not work on other environments due to differing IDs, so we'll need to sort those out when we have a plan to deal with that.

### Screenshots
N/A

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [ ] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

**Review by GeorgeBlackwell — APPROVED** [review]
LGTM!

File: DigitalLearningSolutions.Data/Services/UserService.cs [hunk 1]
```diff
@@ -370,28 +361,23 @@ CentreAnswersData centreAnswersData
...
-            var delegateIds = delegateUsers.Select(d => d.Id).ToArray();
-            userDataService.UpdateDelegateAccountDetails(
+            userDataService.UpdateUserDetails(
                 editDelegateDetailsData.FirstName,
                 editDelegateDetailsData.Surname,
                 editDelegateDetailsData.Email,
-                delegateIds
+                centreAnswersData.JobGroupId,
+                1 // TODO This needs correcting to the correct UserId for the delegate record.
             );
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data/Services/UserService.cs:370]: Comment for slack

File: DigitalLearningSolutions.Data/DataServices/UserDataService/DelegateUserDataService.cs [hunk 2]
```diff
@@ -49,34 +53,10 @@ public partial class UserDataService
...
-                        cd.JobGroupID,
-                        jg.JobGroupName
-                    FROM Candidates AS cd
-                    INNER JOIN Centres AS ct ON ct.CentreID = cd.CentreID
-                    INNER JOIN JobGroups AS jg ON jg.JobGroupID = cd.JobGroupID
+            var users = connection.Query<DelegateUser>(
+                @$"{BaseSelectDelegateQuery}
                     WHERE cd.Active = 1 AND
-                         (cd.CandidateNumber = @username OR cd.EmailAddress = @username OR cd.AliasID = @username)",
+                         (cd.CandidateNumber = @username OR cd.EmailAddress = @username)",
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/DataServices/UserDataService/DelegateUserDataService.cs:59]: We should check this behaviour change in [887](https://softwiretech.atlassian.net/browse/HEEDLS-887)

File: DigitalLearningSolutions.Data/DataServices/UserDataService/DelegateUserDataService.cs [hunk 3]
```diff
@@ -85,33 +65,9 @@ public List<DelegateUser> GetDelegateUsersByUsername(string username)
...
-                        cd.Answer6,
-                        cd.JobGroupID,
-                        jg.JobGroupName
-                    FROM Candidates AS cd
-                    INNER JOIN Centres AS ct ON ct.CentreID = cd.CentreID
-                    INNER JOIN JobGroups AS jg ON jg.JobGroupID = cd.JobGroupID
-                    WHERE cd.CandidateNumber = @username OR cd.EmailAddress = @username OR cd.AliasID = @username",
+            var users = connection.Query<DelegateUser>(
+                @$"{BaseSelectDelegateQuery}
+                    WHERE cd.CandidateNumber = @username OR cd.EmailAddress = @username",
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/DataServices/UserDataService/DelegateUserDataService.cs:70]: Mention on https://softwiretech.atlassian.net/browse/HEEDLS-887

File: DigitalLearningSolutions.Data/DataServices/UserDataService/DelegateUserDataService.cs [hunk 4]
```diff
@@ -193,40 +98,53 @@ public List<DelegateUser> GetUnapprovedDelegateUsersByCentreId(int centreId)
             return users;
         }
 
-        public void UpdateDelegateUsers(
+        public void UpdateUser(
             string firstName,
             string surname,
             string email,
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/DataServices/UserDataService/DelegateUserDataService.cs:104]: Can we rename the param to primaryEmail?

Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data/DataServices/UserDataService/DelegateUserDataService.cs:104]: Also mention on https://softwiretech.atlassian.net/browse/HEEDLS-887

File: DigitalLearningSolutions.Data/DataServices/UserDataService/DelegateUserDataService.cs [hunk 5]
```diff
@@ -193,40 +98,53 @@ public List<DelegateUser> GetUnapprovedDelegateUsersByCentreId(int centreId)
...
+                    userId,
+                    professionalRegNumber,
+                    hasBeenPromptedForPrn,
+                    jobGroupId,
+                }
             );
         }
 
-        public void UpdateDelegateAccountDetails(string firstName, string surname, string email, int[] ids)
+        public void UpdateUserDetails(string firstName, string surname, string email, int jobGroupId, int userId)
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/DataServices/UserDataService/DelegateUserDataService.cs:137]: Mention on https://softwiretech.atlassian.net/browse/HEEDLS-887

File: DigitalLearningSolutions.Data/DataServices/UserDataService/DelegateUserDataService.cs [hunk 6]
```diff
@@ -419,40 +261,6 @@ public void UpdateDelegateLhLoginWarningDismissalStatus(int delegateId, bool sta
             );
         }
 
-        public IEnumerable<DelegateUser> GetDelegateUsersByAliasId(string aliasId)
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/DataServices/UserDataService/DelegateUserDataService.cs:422]: mention on https://softwiretech.atlassian.net/browse/HEEDLS-887

File: DigitalLearningSolutions.Data/Services/UserService.cs [hunk 7]
```diff
@@ -204,15 +207,15 @@ public List<CentreUserDetails> GetUserCentres(AdminUser? adminUser, List<Delegat
...
+                userDataService.UpdateUser(
                     myAccountDetailsData.FirstName,
                     myAccountDetailsData.Surname,
                     myAccountDetailsData.Email,
                     myAccountDetailsData.ProfileImage,
                     myAccountDetailsData.ProfessionalRegistrationNumber,
                     myAccountDetailsData.HasBeenPromptedForPrn,
-                    delegateIds
+                    centreAnswersData.JobGroupId,
+                    1 // TODO This needs correcting to the correct UserId for the delegate record.
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/Services/UserService.cs:218]: Can you stick a ticket reference on this? 887 unless you see another ticket which you're sure should cover this.

Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data/Services/UserService.cs:218]: Same with other todos

**Review by SteveJacksonSoft — COMMENTED** [review]
Worth also commenting on this ticket which features have changed in case they can be tested sensibly - quite a few of them probably can actually.

File: DigitalLearningSolutions.Data.Tests/Services/DelegateUploadFileServiceTests.cs [hunk 8]
```diff
@@ -826,7 +665,7 @@ public void ProcessDelegateTable_counts_updated_correctly()
 
             A.CallTo(() => userDataService.GetDelegateUserByCandidateNumber(delegateId, CentreId))
                 .Returns(candidateNumberDelegate);
-            ACallToUserDataServiceUpdateDelegateDoesNothing();
+            CallsToUserDataServiceUpdatesDoesNothing();
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data.Tests/Services/DelegateUploadFileServiceTests.cs:668]: Something wrong with the grammar here...

File: DigitalLearningSolutions.Data/DataServices/UserDataService/DelegatePromptsDataService.cs [hunk 9]
```diff
@@ -7,7 +7,6 @@ public partial class UserDataService
     {
         public void UpdateDelegateUserCentrePrompts(
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/DataServices/UserDataService/DelegatePromptsDataService.cs:8]: Mention on 887

File: DigitalLearningSolutions.Data/DataServices/UserDataService/AdminUserDataService.cs [hunk 10]
```diff
@@ -58,15 +58,6 @@ ELSE cc.CategoryName
             return user;
         }
 
-        public AdminUser? GetAdminUserByUsername(string username)
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/DataServices/UserDataService/AdminUserDataService.cs:61]: Mention on 887

File: DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests/DelegateUserCardDataServiceTests.cs [hunk 11]
```diff
@@ -56,20 +60,6 @@ public void GetDelegateUserCardById_does_not_match_admin_if_admin_email_address_
             userCard.AdminId.Should().BeNull();
         }
 
-        [Test]
```
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests/DelegateUserCardDataServiceTests.cs:59]: GetDelegateUserCardById_does_not_match_admin_if_password_does_not_match() - If we have a potentially matching admin account, we're going to the Users table to the same record due to the unique emails, so passwords MUST match.

GetDelegateUserCardsByCentreId_does_not_match_admin_if_admin_email_address_is_blank() - Delegate accounts cannot have empty emails any more, therefore the query cannot match to the admin with the blank email.

GetDelegateUserCardsByCentreId_does_not_match_admin_if_password_does_not_match() - Same logic as the GetById method

[the remark this answers is not in the document]
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data.Tests/DataServices/UserDataServiceTests/DelegateUserCardDataServiceTests.cs:59]: Nice, thanks.

**Review by SteveJacksonSoft — APPROVED** [review]
Looks good!

**Pull request merged**